STM/src/test/java/org/jboss/stm/TaxonomyTest.java

```diff
@@ -52,157 +52,6 @@ public void set(int value) throws Exception {
         }
     }
 
-    /**
-     * aborting changes made by a closed nested transaction should not be visible to the parent
-     */
-    public void testIsClosedNestedAbort() throws Exception {
-        AtomicInt ai = new RecoverableContainer<AtomicInt>().enlist(new AtomicIntImpl());
-        AtomicAction outer = new AtomicAction();
-        AtomicAction inner = new AtomicAction();
-
-        ai.set(1); // initialise the shared memory
-        outer.begin(); // start a top level transaction
-        {
-            ai.set(2); // modify state
-            inner.begin(); // start a nested transacction
-            {
-                ai.set(3); // and modify the state
-
-                inner.abort();
-            }
-            outer.commit();
-        }
-
-        // since the child aborted and nesting follows the closed mode the value should still be 2
-        assertEquals(2, ai.get());
-    }
-
-    /**
-     * committing changes made by a closed nested transaction should be visible to the parent
-     */
-    public void testIsClosedNestedCommit() throws Exception {
```

Comment:
As I don't understand all the inner details of the tescases I just wonder if this `testIsClosedNestedCommit` matches in functionality with `testNestedCommit` or if the `testNestedCommit` is a different testcase and the `testIsClosedNestedCommit` was removed for some reason.

Reply:
They are testing different things: the first test runs the nested action in a separate thread whereas the second test runs the nested action in the same thread as the parent. I do have some brief javadoc and java comments above each method that explain what the tests are doing but I will add more text to make that clearer.

Secondly, the first test still exists but I moved it to TaxonomyNestedTest.java and the reason I did that is because the tests for nested actions require the transactional memory to be marked with @Nested (whereas the other tests should not carry this annotation). This change is part of the bug fix for JBTM-3105 that this PR is fixing:

> There are two issues with the nested tests:

> the transactional object is not marked with @Nested
> the second thread that performs a nested update catches Exceptions with fail() in the catch  block. This does not seem to cause the test to fail. The fix is twofold, catch Throwable and assert > failure in the main thread (see PR for the fix)

Thirdly, I also improve coverage so you will see some extra tests.

And finally, I also improved the testability by adding some communication between the parent and child threads so that the parent can check that the child thread is behaving correctly. This change simplifies how the test operates and makes it more resilient since it uses a different approach to inter thread communication (basically it uses message passing between the threads rather than synchronisation primitives which is easier to reason about).

Reply:
+1 thank you for elaboration